### PR TITLE
Week2-React homework-Osama Bazbooz

### PR DESCRIPTION
**Pull request by obazbooz**

### CODE REVIEW

**Comment by crevulus** [comment]
Screenshot 2022-04-27 at 09 25 52
 
@obazbooz You are missing a veeeeeery important package from your dependencies! I can't run the app without it.

**Comment by obazbooz** [comment]
Dear @crevulus ,

Sorry about that bug, I updated the code. Hopefully the bug solved.
Thanks for your time.

Best regards,
Osama

**Comment by crevulus** [comment]
Hi @obazbooz ,

Your latest commits did not fix the bug. Please check your **dependencies** in your **package.json**. You are missing `react-router-dom`, which is the package that handles routing.

File: week2/project/ecommerce/src/App.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    })();
+  }, []);
+
+  return (
+    <div className="App">
+      <h1 className="mainTitle">MEDIA OCEAN</h1>
+      {isLoading && !isFail ? (
+        <div>
+          <img
+            className="loadingImg"
```
Comment (crevulus) [inline week2/project/ecommerce/src/App.js:59]: You are getting a warning in your console/VSCode: 
```
Redundant alt attribute. Screen-readers already announce `img` tags as an image. You don’t need to use the words `image`, `photo,` or `picture` (or any specified custom words) in the alt prop.
```

🟠

File: week2/project/ecommerce/src/components/categories.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { useState } from 'react';
+
+function Categories({ setSelectedCategoryHandler, allCategories }) {
+  const [active, setActive] = useState(allCategories[null]);
+  const [flag, setFlag] = useState(false);
+
+  const clickedAgainFlag = (state) => {
+    setFlag(state);
+  };
+  let categoryName = '';
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:10]: This is not being used; you can delete it. 🟢

File: week2/project/ecommerce/src/components/categories.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+              className="categoryBtn"
+              style={{
+                opacity: active === category && !flag ? 1 : 0.3,
+                boxShadow:
+                  active === category && !flag ? '0px 0px 6px #661E0C' : '',
+              }}
+              onClick={() => {
+                setActive(category);
+                if (active === category && flag === false) {
+                  clickedAgainFlag(true);
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:28]: Do you need this "clicked again flag"? You've actually got the secret in your code already! In your click handler you can do `if active === category` and pass in `'All'` if it's true or just `category` if it's false. 🔴

File: week2/project/ecommerce/src/components/categories.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          (
+            <button
+              key={index}
+              className="categoryBtn"
+              style={{
+                opacity: active === category && !flag ? 1 : 0.3,
+                boxShadow:
+                  active === category && !flag ? '0px 0px 6px #661E0C' : '',
+              }}
+              onClick={() => {
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:25]: Try not to put large onclick functions inline. Separate them into their own function. That makes it much easier to read and maintain. 🟢

File: week2/project/ecommerce/src/components/categories.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  const clickedAgainFlag = (state) => {
+    setFlag(state);
+  };
+  let categoryName = '';
+  return (
+    <div className="categories">
+      {allCategories.map(
+        (category, index) => (
+          (categoryName = category.slice(6)),
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:15]: This syntax is quite... odd? It works so it's okay, but you're untlikely to see it anywhere else, I think! it would be better to do `(category, index) => {const categoryName = catgory.slice(6); return (<button>...</button>)}` 🟢

File: week2/project/ecommerce/src/components/productItem.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { Link } from 'react-router-dom';
+import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
+import ProductItemDetails from './productItemDetails.js';
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItem.js:3]: Remove imports you're not using! 🟢

File: week2/project/ecommerce/src/components/productItem.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+function ProductsItem({ product: { id, title, description, image } }) {
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItem.js:5]: Very nice destructuring!

File: week2/project/ecommerce/src/components/productItemDetails.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        console.log(error);
+        setIsFail(true);
+      }
+    })();
+  }, []);
+
+  return (
+    <div className="productDetailsContainer">
+      <h1 className="mainTitle">MEDIA OCEAN</h1>
+      {isLoading && !isFail ? (
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItemDetails.js:32]: Good error handling again!

File: week2/project/ecommerce/src/components/productItemDetails.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        <img
+          className="loadingImg"
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItemDetails.js:34]: You are getting the same error here about alt tags that you got in App.js. 🟢

File: week2/project/ecommerce/src/components/productItemDetails.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+          src="/isloading.gif"
+          alt="Image not found!"
+        />
+      ) : (
+        <div>
+          <img className="detailsImg" src={prodoctDetails.image} />
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItemDetails.js:40]: Missing an alt tag - they're very important! 🟠

File: week2/project/ecommerce/src/components/productItemDetails.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+              {prodoctDetails.description}
+            </p>
+            <p>
+              <span>Product Rating: </span>
+              {prodoctDetails.rating.rate}
+            </p>
+          </div>
+        </div>
+      )}
+      {isFail ? (
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/productItemDetails.js:69]: You can shorten this to one line with the `&&` operator, rather than a ternary operator. 🟢

File: week2/project/ecommerce/src/index.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import { createRoot } from 'react-dom/client';
+import './index.css';
+import App from './App';
+import reportWebVitals from './reportWebVitals';
+import ProductItemDetails from './components/productItemDetails.js';
+
+const container = document.getElementById('root');
+const root = createRoot(container);
+root.render(
+  <Router>
```
Comment (crevulus) [inline week2/project/ecommerce/src/index.js:12]: This is the wrong place to put your routing logic. The `App` component should be the gateway to your whole application. There shouldn't be a conditional option where someone either goes to your `App` or somewhere else. This is causing re-rendering on your root DOM element, which can cause problems in React. Look again at the instructions from react-router: the `BrowserRouter` wraps the `App`, and then in `App` you add your routing logic: https://reactrouter.com/docs/en/v6/getting-started/tutorial#add-some-routes 🔴

File: week2/project/ecommerce/src/App.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+      setSelectedCategory(category);
+    }
+  };
+
+  useEffect(() => {
+    (async () => {
+      try {
+        const categoriesApiResponse = await fetch(categoriesApiUrl);
+        const categoriesApiData = await categoriesApiResponse.json();
+        setTimeout(() => {
```
Comment (crevulus) [inline week2/project/ecommerce/src/App.js:28]: What is the timeout for here? timeouts should be avoided whenever possible (unless you ahve something on the UI that specifically requires a set time period) 🔴

File: week2/project/ecommerce/src/App.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+          setAllCategories(categoriesApiData);
+          setIsLoading(false);
```
Comment (crevulus) [inline week2/project/ecommerce/src/App.js:30]: When do you set `isLoading` to true? The purpose of the loading image is to demonstrate that data is loading, right? So at the start of your call you should set the state to `true` - and then keep the `setIsLoading(false);` like you have now. Your current set up works, but this is a good practice for you to get familiar with. 🟢

**Review by crevulus — CHANGES_REQUESTED** [review]
Hi Osama,

Thanks for this PR. You've made a great start! It's difficult to tell exactly how your app works seeing as there is still a bug that you need to fix in order for it to work. I've tried to leave as much feedback as I can. Let me know if you have any questions.

**Comment by obazbooz** [comment]
Dear @crevulus  ,

I have worked on your valuable feedback and fixed my code. I will be happy to hear your feedback about my fixed code.

Thanks a lot for all the support and help.
Best regards,
Osama

File: week2/project/ecommerce/src/components/categories.js (hunk 3 above)
Comment (obazbooz) [reply week2/project/ecommerce/src/components/categories.js]: I really need it because without that flag I am not able to deselect the category if I click it again, and also I had issue without it which is when I switch between categories I have to click twice and the category btn to get only the products which are related to the selected category .

File: week2/project/ecommerce/src/App.js (hunk 13 above)
Comment (obazbooz) [reply week2/project/ecommerce/src/App.js]: I have just use it to slow down the fetching a bit to show the loading img.
I removed them all.

File: week2/project/ecommerce/src/components/shop.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+  const setSelectedCategoryHandler = (category) => {
+    if (category === 'All') {
+      setSelectedCategory('All');
+    } else {
+      setSelectedCategory(category);
+    }
+  };
+
+  useEffect(() => {
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/shop.js:22]: You are missing a dependency array in your useEffect! It's very important that you include one unless you want to trigger lots of re-renders and lots of fetch calls. https://blog.bitsrc.io/understanding-dependencies-in-useeffect-7afd4df37c96 🔴

File: week2/project/ecommerce/src/components/categories.js (hunk 3 above)
Comment (crevulus) [reply week2/project/ecommerce/src/components/categories.js]: I sent you a video in Slack to view how you can fix this.

File: week2/project/ecommerce/src/components/categories.js [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  };
+
+  const renderCategoryBtn = (category, index) => {
+    return (
+      <button
+        key={index}
+        className="categoryBtn"
+        style={{
+          opacity: active === category && !flag ? 1 : 0.3,
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:28]: Use css or inline styles, not both. 🔴

File: week2/project/ecommerce/src/components/categories.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    if (active === category && flag === false) {
+      clickedAgainFlag(true);
+      setSelectedCategoryHandler('All');
+    } else {
+      clickedAgainFlag(false);
+      setSelectedCategoryHandler(category);
+    }
+  };
+
+  const renderCategoryBtn = (category, index) => {
```
Comment (crevulus) [inline week2/project/ecommerce/src/components/categories.js:22]: Why does this need to be a separate function? You can simply return jsx in your `map` function down below. 🔴

File: week2/project/ecommerce/src/App.css [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+.mainTitle {
+  font-family: 'Monoton', cursive;
+  font-family: 'Rubik Moonrocks', cursive;
+}
+
+/************************Categories***********************************/
+
+.categories {
```
Comment (crevulus) [inline week2/project/ecommerce/src/App.css:12]: Check your styling in mobile! 🔴

File: week2/project/ecommerce/src/components/categories.js (hunk 16 above)
Comment (obazbooz) [reply week2/project/ecommerce/src/components/categories.js]: Sorry I  could not figure it out how to have all the styling in the css file if I have conditional styling required. Can u help me with a hint to do that ?

**Comment by obazbooz** [comment]
Dear @crevulus ,

Thanks a lot for your feedback. I have fixed the issues.

Best regards,
Osama

File: week2/project/ecommerce/src/components/categories.js (hunk 16 above)
Comment (crevulus) [reply week2/project/ecommerce/src/components/categories.js]: You can do it in `className`! Create two classes with the styling that you want and return them conditionally. So in this case you could make a `selected` class and an `unselected` class, for example. Then in `className`, do the same conditional logic here but use it to apply/remove the classes you need, i.e. className={` categoryBtn ${active === category ? "selected" : "unselected"}`}

**Comment by crevulus** [comment]
Reviewed and approved, but please check out my comment about the conditional classes!

**Review by crevulus — APPROVED** [review]

**Comment by obazbooz** [comment]
Dear @crevulus ,

I have fixed the inline styling issue.
Thanks a lot for all ur support and help.

Best regards,
Osama
